### ISSUE TITLE
TypeScript: what connection type to use?

### ISSUE THREAD

**Issue by nodefish** [gajus/slonik#228]
I'm a bit confused about what type to use for a normal connection.

The following causes the compiler to complain with `"Cannot invoke an object which is possibly 'undefined'"` on `conn.transaction`. This is because `DatabaseConnectionType` is `Partial`.

```ts
await pool.connect(async (conn: DatabaseConnectionType) => {
        await conn.transaction(async (tx) => {
            ...
        }
}
```
It's fixed by using `DatabasePoolConnectionType` or `DatabaseTransactionConnectionType`, but I don't understand how the former differs semantically from `DatabaseConnectionType` (I don't get the conceptual difference in plain English) and the latter is wrong.

**Comment by nodefish** [comment]
Additionally, it's often desirable to have a function or method that accepts a connection argument, and it'd be nice if it were polymorphic and could accept normal or transaction connections. 

Example:
```ts
async function doSomethingWithAConnection(conn: DatabasePoolConnectionType, ...) {
    ...
}
```

This wouldn't compile if called with a `DatabaseTransactionConnectionType` as the argument.

**Comment by pdlug** [comment]
I'm having the same issue. I solved it in my code by adding `export type SlonikDBConnection = DatabasePoolConnectionType | DatabaseTransactionConnectionType;` to one of my util libraries and then just using `SlonikDBConnection`
